language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -315,3 +362,34 @@ onHead :: (a -> a) -> [a] -> [a]
 onHead f = \case
     [] -> []
     x:xs -> f x:xs
+
+onLast :: (a -> a) -> [a] -> [a]
+onLast f = \case
+  [] -> []
+  [l] -> [f l]
+  x : xs -> x : onLast f xs
+
+-- A type such as
+--   data Q = C { x: Int, y: Text }| G { z: Bool }
+-- has a DAML-LF representation like,
+--   record Q.C = { x: Int, y: String }
+--   record Q.G = { z: Bool }
+--   variant Q = C Q.C | G Q.G
+-- Use this function to get the definitions of 'Q.C' and 'Q.G' given
+-- 'Q'.
+assocDefDataTypes :: Module -> T.Text -> [((T.Text, T.Text), DefDataType)]
+assocDefDataTypes m c =
+  [((sup, sub ), def) | def <- defDataTypes m
+    , [sup, sub] <- [unTypeConName (dataTypeCon def)], sup == c]
+
+-- Extract the serialization code associated with a data type
+-- definition.
+serFromDef :: ([T.Text] -> c)
```

Comment:
I'm not a big fan of this layout of type signatures. If you change the name of the function (to a different length) you either get diff noise or ugly code.

Reply:
I'm generally not either. I'll go linear on it.

Reply:
When it gets too long, I usually write
```
f ::
  A ->
  B ->
  C
```
This is maybe not the most pretty but does not have any of the issues I mentioned above. Others would write
```
f
  :: A
  -> B
  -> C
```
but I don't like that too much because it gets diff-noisy if you want to add a new first parameter.

Reply:
Moved to the `where` clause, no signature since no longer top-level.